src/Parser/Core/Modules/Spells/BFA/AzeriteTraits/SecondaryStatProcTrait.js

```diff
@@ -0,0 +1,143 @@
+import React from 'react';
+
+import { formatNumber, formatPercentage } from 'common/format';
+import { calculateAzeriteEffects } from 'common/stats';
+import Analyzer from 'Parser/Core/Analyzer';
+import TraitStatisticBox, { STATISTIC_ORDER } from 'Interface/Others/TraitStatisticBox';
+import StatTracker from 'Parser/Core/Modules/StatTracker';
+
+const CRIT = 'crit';
+const VERSA = 'versa';
+const MASTERY = 'mastery';
+const HASTE = 'haste';
+
+const STATS = {
+  [CRIT]: {
+    name: 'Crit',
+    percentageFn: 'critPercentage',
+  },
+  [VERSA]: {
+    name: 'Versatility',
+    percentageFn: 'versatilityPercentage',
+  },
+  [MASTERY]: {
+    name: 'Mastery',
+    percentageFn: 'masteryPercentage',
+  },
+  [HASTE]: {
+    name: 'Haste',
+    percentageFn: 'hastePercentage',
+  },
+};
+
+class SecondaryStatProcTrait extends Analyzer {
+  static CRIT = CRIT;
+  static VERSA = VERSA;
+  static MASTERY = MASTERY;
+  static HASTE = HASTE;
+
+  static dependencies = {
+    statTracker: StatTracker,
+  };
+
+  // info about trait passed in by subclass
+  traitId;
+  traitName;
+  traitIcon;
+  buffId;
+  stats;
+
+  // store trait ranks from combatant
+  traitRanks;
+
+  constructor(traitId, traitName, traitIcon, buffId, stats, ...args) {
+    super(...args);
+
+    this.traitId = traitId;
+    this.traitName = traitName;
+    this.traitIcon = traitIcon;
+    this.buffId = buffId;
+    this.stats = stats;
+
+    this.active = this.selectedCombatant.hasTrait(this.traitId);
+    if (!this.active) {
+      return;
+    }
+
+    this.traitRanks = this.selectedCombatant.traitRanks(this.traitId);
+
+    // sanity check to double check the subclass is implemented correctly matching the data we have about the trait
+    if (this.stats.length !== this.statBuffs.length) {
+      throw new Error("Mismatch between stat buffs specified in azerite trait effects and passed in from subclass");
+    }
+  }
+
+  get statBuffs() {
+    return Object.values(this.traitRanks).reduce((stats, rank) => {
+      calculateAzeriteEffects(this.traitId, rank).forEach((statBuff, idx) => {
+        stats[idx] += statBuff;
+      });
+
+      return stats;
+    }, calculateAzeriteEffects(this.traitId, 0).map(() => 0));
+  }
+
+  get uptime() {
+    return this.selectedCombatant.getBuffUptime(this.buffId) / this.owner.fightDuration;
+  }
+
+  /**
+   * to make the tooltip reflect multiple ranks of the trait we use a bit of a hacky way to figure out which ilvl would result in displaying the sum of the bonus damage
+   */
+  determineIlvlForSum() {
```

Comment:
I took this code from my other PR, it's a bit funky and off slightly but I don't see another way to come to an ilvl that when passed into the wowhead tooltip will be close to the sum of the traits effects.